File: leetcode-swift/lc_013.swift
//
// LeetCode
// 13. Roman to Integer
// https://leetcode.com/problems/roman-to-integer/
//


class Solution {
    private enum RomanValue {
        static let M = 1000
        static let D = 500
        static let C = 100
        static let L = 50
        static let X = 10
        static let V = 5
        static let I = 1
    }

    func romanToInt(_ s: String) -> Int {
        var last = RomanValue.M
        var num: Int
        var result = 0

        for c in s {
            switch c {
            case "M":
                num = RomanValue.M
            case "D":
                num = RomanValue.D
            case "C":
                num = RomanValue.C
            case "L":
                num = RomanValue.L
            case "X":
                num = RomanValue.X
            case "V":
                num = RomanValue.V
            case "I":
                num = RomanValue.I
            default:
                return 0
            }

            result += num
            if num > last {
                result -= 2 * last
            }
            last = num
        }

        return result
    }
}

File: leetcode-go/lc_013.go
//
// LeetCode
// 13. Roman to Integer
// https://leetcode.com/problems/roman-to-integer/
//


func romanToInt(s string) int {
    romanVal := map[byte]int{
        'M': 1000,
        'D': 500,
        'C': 100,
        'L': 50,
        'X': 10,
        'V': 5,
        'I': 1,
    }

    last := romanVal['M']
    result := 0

    for i := 0; i < len(s); i++ {
        if val, ok := romanVal[s[i]]; ok {
            result += val

            if val > last {
                result -= 2 * last
            }

            last = val
        } else {
            return 0
        }
    }

    return result
}
